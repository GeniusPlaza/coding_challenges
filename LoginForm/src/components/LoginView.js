import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  TextInput,
  Button,
  View,
} from 'react-native';

import {loginRequest} from '../Requests';

class LoginView extends React.Component {
  _username = "";
  _password = "";

  onSetUsername = (text) => {
    this._username = text;
  }
  
  onSetPassword = (text) => {
    this._password = text;
  }

  onLoginRequest = (username, password) => {
    if (username.length == 0)
      this.props.onFail("Missing username");
    else if (password.length == 0)
      this.props.onFail("Missing password");
    else
    {
      // There should be a way to init the _loginRequest in the constructur. Not sure how to do it yet.
      let _loginRequest = this.props.loginRequest != null ? this.props.loginRequest : loginRequest;

      // loginRequest.Login(_username, _password);
      _loginRequest(
        username, 
        password, 
        ()=>{
          this.props.onSuccess();
        },
        (error)=>{
          this.props.onFail(error);
        }
      );
    }
  };

//<Text style={styles.subTitle}>Username</Text>
//<Text style={styles.subTitle}>Password</Text>

  render() {
    return (
      <View style={styles.container}>
        <Text style={styles.title}>Login</Text>
        <View style={styles.boxContainer}>
          <TextInput 
            placeholder = "username"
            style = {styles.textInput}
            onChangeText={text => this.onSetUsername(text)}
          />
          <TextInput 
            placeholder = "password"
            style = {styles.textInput}
            onChangeText={text => this.onSetPassword(text)}
          />
          <View style={styles.buttonContainer}>
            <Button style={styles.button} title="Login" onPress={()=>this.onLoginRequest(this._username, this._password)}/>
          </View>
        </View>
      </View>
    );
  }
};

export default LoginView;

const styles = StyleSheet.create({
  title:{
    fontSize: 40,
    color: '#777',
    textAlign: 'center',
    marginTop: 20,
    marginBottom: 100
  },
  subTitle:{
    fontSize: 20,
    color: 'black',
    marginTop: 10
  },
  container: {
    flex: 1,
    backgroundColor: '#ddd'
  },
  boxContainer:{
    //flex: 1,
    backgroundColor: '#fff',
    height: 270,
    padding: 30
  },
  textInput: {
    fontSize: 20,
    borderColor: 'black',
    borderWidth: 1,
    padding: 10,
    marginBottom: 20
  },
  buttonContainer:{
    marginTop: 20,
    marginLeft: 30,
    marginRight: 30,
    borderColor: 'black',
  },
  // none of the button styles are working
  button:{
    fontSize: 40,
    height: 40
  }
});
