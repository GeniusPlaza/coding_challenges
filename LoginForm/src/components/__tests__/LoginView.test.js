import React from 'react';
import renderer from 'react-test-renderer';
import LoginView from "../LoginView";
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('All LoginView Tests', () => {

    it('renders correctly', () => {
        const tree = renderer.create(<LoginView/>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('starts a login request', () => {
        var loginRequestMock = jest.fn((username, pass, onSuccess, onFail)=>{});
        const wrapper = shallow(<LoginView 
            loginRequest = {loginRequestMock}
        />);
        const instance = wrapper.instance();
        instance.onLoginRequest("user", "pass");

        expect(loginRequestMock.mock.calls.length).toBe(1);
    });

    it('handles a succesful request', () => {
        var handledSuccess = false;
        var handledFail = false;
        var successLoginRequestMock = jest.fn((username, pass, onSuccess, onFail)=>{onSuccess()});
        const wrapper = shallow(<LoginView 
            loginRequest = {successLoginRequestMock}
            onSuccess = {()=>handledSuccess=true}
            onFail = {()=>handledFail=true}
        />);
        const instance = wrapper.instance();

        instance.onLoginRequest("user", "pass");
        
        expect(handledSuccess).toBe(true);
        expect(handledFail).toBe(false);
    });

    it('handles a failed request', () => {
        // this is the 3rd time I repeat the same Arrange code. I need to find a way to reduce duplication.
        var handledSuccess = false;
        var handledFail = false;
        var failLoginRequestMock = jest.fn((username, pass, onSuccess, onFail)=>{onFail()});
        const wrapper = shallow(<LoginView 
            loginRequest = {failLoginRequestMock}
            onSuccess = {()=>handledSuccess=true}
            onFail = {()=>handledFail=true}
        />);
        const instance = wrapper.instance();

        instance.onLoginRequest("user", "pass");

        expect(handledFail).toBe(true);
        expect(handledSuccess).toBe(false);
    });
});