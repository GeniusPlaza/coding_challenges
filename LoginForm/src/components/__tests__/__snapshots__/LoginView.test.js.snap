// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All LoginView Tests renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ddd",
      "flex": 1,
    }
  }
>
  <Text
    style={
      Object {
        "color": "black",
        "fontSize": 30,
      }
    }
  >
    Login Screen
  </Text>
  <Text
    style={
      Object {
        "color": "black",
        "fontSize": 20,
      }
    }
  >
    Username
  </Text>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="username"
    rejectResponderTermination={true}
    style={
      Object {
        "borderColor": "black",
        "borderWidth": 1,
        "fontSize": 20,
        "padding": 10,
      }
    }
    underlineColorAndroid="transparent"
  />
  <Text
    style={
      Object {
        "color": "black",
        "fontSize": 20,
      }
    }
  >
    Password
  </Text>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="password"
    rejectResponderTermination={true}
    style={
      Object {
        "borderColor": "black",
        "borderWidth": 1,
        "fontSize": 20,
        "padding": 10,
      }
    }
    underlineColorAndroid="transparent"
  />
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Login
      </Text>
    </View>
  </View>
</View>
`;
